File: tests/python/test_first_order.py
from __future__ import annotations

import numpy as np

import optlib


def test_minimize_rosenbrock_returns_result_and_trajectory() -> None:
    result = optlib.MinimizeRosenbrock(
        np.array([1.2, 1.2]),
        method="adam",
        learning_rate=1e-2,
        max_iter=30000,
        gradient_tolerance=1e-4,
        step_tolerance=0.0,
        function_tolerance=0.0,
    )

    assert result["value"] < 1e-4
    assert result["iterations"] > 0
    assert result["trajectory"]["x"].shape[1] == 2
    assert result["trajectory"]["f"].shape[0] == result["trajectory"]["x"].shape[0]


def test_general_minimize_accepts_python_callables() -> None:
    target = np.array([1.0, -2.0, 0.5])

    def value(values: np.ndarray) -> float:
        shifted = values - target
        return float(np.dot(shifted, shifted))

    def gradient(values: np.ndarray) -> np.ndarray:
        return 2.0 * (values - target)

    result = optlib.Minimize(
        value,
        gradient,
        np.array([4.0, 4.0, 4.0]),
        method="gradient_descent",
        learning_rate=0.1,
        max_iter=1000,
        gradient_tolerance=1e-8,
        step_tolerance=0.0,
        function_tolerance=0.0,
    )

    assert result["converged"]
    np.testing.assert_allclose(result["x"], target, atol=1e-6)
    assert result["value"] < 1e-12

File: tests/cpp/TestFirstOrder.cpp
#include <gtest/gtest.h>

#include <optlib/core/optimizers/FirstOrder.hpp>

namespace {

optlib::FirstOrderConfig ConfigFor(optlib::FirstOrderMethod method)
{
    optlib::FirstOrderConfig config;
    config.Criteria.MaxIterations = 20000;
    config.Criteria.GradientTolerance = 1e-3;
    config.Criteria.StepTolerance = 0.0;
    config.Criteria.FunctionTolerance = 0.0;
    config.StoreTrajectory = true;

    switch (method) {
    case optlib::FirstOrderMethod::GradientDescent:
        config.LearningRate = 1e-3;
        break;
    case optlib::FirstOrderMethod::HeavyBall:
        config.LearningRate = 8e-4;
        config.Momentum = 0.8;
        break;
    case optlib::FirstOrderMethod::Nesterov:
        config.LearningRate = 8e-4;
        config.Momentum = 0.8;
        break;
    case optlib::FirstOrderMethod::Adam:
        config.LearningRate = 2e-2;
        break;
    case optlib::FirstOrderMethod::RMSProp:
        config.LearningRate = 2e-4;
        config.Beta2 = 0.999;
        break;
    case optlib::FirstOrderMethod::Adagrad:
        config.LearningRate = 5e-2;
        break;
    }
    return config;
}

} // namespace

TEST(FirstOrderTest, MethodsReduceRosenbrock2D)
{
    const optlib::Vector initialX{-1.2, 1.0};
    for (const optlib::FirstOrderMethod method :
         {optlib::FirstOrderMethod::GradientDescent, optlib::FirstOrderMethod::HeavyBall,
          optlib::FirstOrderMethod::Nesterov, optlib::FirstOrderMethod::Adam,
          optlib::FirstOrderMethod::RMSProp, optlib::FirstOrderMethod::Adagrad}) {
        optlib::FirstOrderConfig config = ConfigFor(method);
        const optlib::OptimizeResult result =
            optlib::MinimizeRosenbrock(initialX.Span(), method, config);
        EXPECT_LT(result.Value, 1e-2);
        EXPECT_LT(result.GradientNorm, 0.25);
        EXPECT_GT(result.Path.Size(), 1U);
        EXPECT_EQ(result.Path.Dimension(), initialX.Size());
    }
}

TEST(FirstOrderTest, AdamConvergesOnRosenbrockND)
{
    optlib::Vector initialX{-1.2, 1.0, -1.2, 1.0, -1.2};
    optlib::FirstOrderConfig config = ConfigFor(optlib::FirstOrderMethod::Adam);
    config.Criteria.MaxIterations = 30000;
    const optlib::OptimizeResult result =
        optlib::MinimizeRosenbrock(initialX.Span(), optlib::FirstOrderMethod::Adam, config);
    EXPECT_LT(result.Value, 1e-2);
    EXPECT_LT(result.GradientNorm, 0.25);
}
